src/types/boxed.rs

```diff
@@ -0,0 +1,343 @@
+/// or one of the other types that provide interior mutability.
+///
+/// ## `Deref` behavior
+///
+/// `JsBox<T>` automatically dereferences to `T` (via the `Deref` trait), so
+/// you can call `T`'s method on a value of type `JsBox<T>`.
+///
+/// ```rust
+/// # use neon::prelude::*;
+/// # fn my_neon_function(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+/// let vec: Handle<JsBox<Vec<_>>> = cx.boxed(vec![1, 2, 3]);
+///
+/// println!("Length: {}", vec.len());
+/// # Ok(cx.undefined())
+/// # }
+/// ```
+///
+/// ## Examples
+///
+/// Passing some immutable data between Rust and JavaScript.
+///
+/// ```rust
+/// # use neon::prelude::*;
+/// # use std::path::{Path, PathBuf};
+/// fn create_path(mut cx: FunctionContext) -> JsResult<JsBox<PathBuf>> {
+///     let path = cx.argument::<JsString>(0)?.value(&mut cx);
+///     let path = Path::new(&path).to_path_buf();
+///
+///     Ok(cx.boxed(path))
+/// }
+///
+/// fn print_path(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+///     let path = cx.argument::<JsBox<PathBuf>>(0)?;
+///
+///     println!("{}", path.display());
+///
+///     Ok(cx.undefined())
+/// }
+/// ```
+///
+/// Passing a user defined struct wrapped in a `RefCell` for mutability. This
+/// pattern is useful for creating classes in JavaScript.
+///
+/// ```rust
+/// # use neon::prelude::*;
+/// # use std::cell::RefCell;
+///
+/// type BoxedPerson = JsBox<RefCell<Person>>;
+///
+/// struct Person {
+///      name: String,
+/// }
+///
+/// impl Finalize for Person {}
+///
+/// impl Person {
+///     pub fn new(name: String) -> Self {
+///         Person { name }
+///     }
+/// 
+///     pub fn set_name(&mut self, name: String) {
+///         self.name = name;
+///     }
+/// 
+///     pub fn greet(&self) -> String {
+///         format!("Hello, {}!", self.name)
+///     }
+/// }
+/// 
+/// fn person_new(mut cx: FunctionContext) -> JsResult<BoxedPerson> {
+///     let name = cx.argument::<JsString>(0)?.value(&mut cx);
+///     let person = RefCell::new(Person::new(name));
+/// 
+///     Ok(cx.boxed(person))
+/// }
+/// 
+/// fn person_set_name(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+///     let person = cx.argument::<BoxedPerson>(0)?;
+///     let mut person = person.borrow_mut();
+///     let name = cx.argument::<JsString>(1)?.value(&mut cx);
+/// 
+///     person.set_name(name);
+/// 
+///     Ok(cx.undefined())
+/// }
+/// 
+/// fn person_greet(mut cx: FunctionContext) -> JsResult<JsString> {
+///     let person = cx.argument::<BoxedPerson>(0)?;
+///     let person = person.borrow();
+///     let greeting = person.greet();
+/// 
+///     Ok(cx.string(greeting))
+/// }
+pub struct JsBox<T: Send + 'static> {
+    local: raw::Local,
+    // `JsBox` cannot verify the lifetime. Store a raw pointer to force uses
+    // to be marked unsafe. In practice, it can be treated as `'static` but
+    // should only be exposed as part of a `Handle` tied to a `Context` lifetime.
+    internal: *const T,
```

Comment:
I can add a comment. It's a pointer to the user's data.

Reply:
I think the things that confused me were:

1. It's the only `JsFoo` type that isn't just a zero-cost wrapper of `raw::Local`.
2. I didn't understand that this is just a cached copy of the extracted and downcast underlying pointer, i.e. just an optimization. It would be helpful to call that out.
3. The fact that it's wider than other `JsFoo` types means it can't be used as a `This` type.

I think if we just document this with internal comments that would be helpful.

Reply:
Per slack discussion, I will be removing the cached pointer optimization. It will be replaced by a `PhantomData`. Any preference on name? I usually go with something generic like `_phantom` or `_marker`.

Reply:
After further thought, the cached pointer is necessary to facilitate `Deref` without needing an `Env`. I will add additional documentation.